src/test/clojure/clara/performance/test_compilation.clj

```diff
@@ -0,0 +1,154 @@
+                                                    :constraints ['(= ?binding-l this)]}]
+                                             :rhs `(r/insert! (with-meta ~(set (repeatedly 10 #(rand-nth (range 100))))
+                                                                         {:type ~(keyword next-fact)})))]
+                            [next-fact production])))]
+    (into {}
+          (for [combo (partition-all 2 (shuffle seed-syms))]
+            (apply template combo)))))
+
+(defn generate-collection-productions
+  [seed-syms]
+  (into {}
+        (for [seed-sym seed-syms
+              :let [next-fact (symbol (str seed-sym "prime"))
+                    production (assoc base-production
+                                 :lhs [{:accumulator '(clara.rules.accumulators/all)
+                                        :from {:type (keyword seed-sym),
+                                               :constraints [`(filter-fn ~'this)]}
+                                        :result-binding :?binding}]
+                                 :rhs `(r/insert! (with-meta ~'?binding
+                                                             {:type ~(keyword next-fact)})))]]
+          [next-fact production])))
+
+(defn generate-queries
+  [seed-syms]
+  (into {}
+        (for [seed-sym seed-syms
+              :let [production (assoc base-production
+                                 :lhs [{:type [seed-sym]
+                                        :constraints []
+                                        :fact-binding :?binding}]
+                                 :params #{})]]
+          [seed-sym production])))
+
+(defn generate-rules-and-opts
+  [num-starter-facts]
+  (let [starter-seeds (repeatedly num-starter-facts gensym)]
+    (-> (reduce (fn [[seeds prods] next-fn]
+                  (let [new-productions-and-facts (next-fn seeds)]
+                    [(keys new-productions-and-facts)
+                     (concat prods
+                             (vals new-productions-and-facts))]))
+                [starter-seeds []]
+                [generate-filter-productions
+                 generate-compose-productions
+                 generate-collection-productions
+                 generate-queries])
+        second
+        vector
+        (conj :cache false))))
+
+(deftest compilation-performance-test
+  (let [rules (generate-rules-and-opts 500)]
+    (testing "Session creation performance"
+      (run-performance-test
+        {:description "Generated Session Compilation"
+         :func #(com/mk-session rules)
+         :iterations 50
+         :mean-assertion (partial > 5000)}))
+
+    (let [session (com/mk-session rules)
+          os (ByteArrayOutputStream.)]
+      (testing "Session serialization performance"
+        (run-performance-test
+          {:description "Generated Session serialization"
+           :func #(.serialize (dura/create-session-serializer (ByteArrayOutputStream.))
```

Comment:
You shouldn't interact with the API via the impl detail that protocols may expose interfaces.
This should be called via the protocol fn var.

```clj
[clara.rules.durability :as dur]
...

(dur/serialize ...)
```

Also, I'd recommend not calling `[clara.rules.durability.fressian :as dura]` since that is confusing.
Perhaps, `[clara.rules.durability.fressian :as fres]` or something